### ISSUE TITLE
enum map key deserialization fails for some json property orders

### ISSUE THREAD

**Issue by saizcorbe** [FasterXML/jackson-databind#1183]
Currently i have a generics polymorphic schema:
Superclass:

```java
@JsonTypeInfo(
        use=JsonTypeInfo.Id.NAME,
        include=JsonTypeInfo.As.PROPERTY,
        property="type",
        visible=false
)
@JsonSubTypes({
        @JsonSubTypes.Type(value = EnumKeysContainer.class, name = "enumKey"),
        @JsonSubTypes.Type(value = StringKeysContainer.class, name = "stringKey")
})
public abstract class GenericContainer<T> {

    private T variableDescription;

    public T getVariableDescription() {
        return variableDescription;
    }

    public void setVariableDescription(T variableDescription) {
        this.variableDescription = variableDescription;
    }
}
```

Problematic Subclass:
```java
public class EnumKeysContainer extends GenericContainer<Map<EnumValuesTest, String>>{
}
```

Enum keys:

```java
public enum EnumValuesTest {
	ENUM_VALUE1,
	ENUM_VALUE2
}
```

Similar class without enum keys:

```java
public class StringKeysContainer extends GenericContainer<Map<String, String>> {
}
```

If i try to deserialize an instance of EnumKeyContainer from this json it fails when trying to deserialize the enum key:

```java
{
  "variableDescription": {"ENUM_VALUE1": "fb"},
  "type": "enumKey"
}
```

The error message is the following:
`com.fasterxml.jackson.databind.exc.InvalidFormatException: Can not construct Map key of type EnumValuesTest from String "ENUM_VALUE1": 
 at [Source: {
  "keysMap": {"ENUM_VALUE1": "fb"},
  "type": "enumKey"
}; line: 2, column: 35] (through reference chain: EnumKeysContainer["keysMap"])
`

but it successfully deserialize the same json with the properties in a different order

```json
{
  "type": "enumKey",
  "variableDescription": {"ENUM_VALUE1": "fb"}
}
```

There are no problems on deserializing StringKeysContainer.
If I try to define a @JsonCreator for the enum

```java
@JsonCreator 
public static EnumValuesTest fromValue(String value)`
it receives the type property value (i.e. 'enumKey') instead of the expected 'ENUM_VALUE1'.  If type property is the first, the @JsonCreator method receives the right token
```

EDIT: Using jackson 2.6.3 and dropwizard 0.9.2

**Comment by cowtowncoder** [comment]
Thank you for indicating the version number.

At this point it is highly likely that the last 2.6.x release is 2.6.6 (released just today), so it would be useful to know if this can be reproduced with 2.7.3.

**Comment by cowtowncoder** [comment]
Unfortunately I can not reproduce this either with 2.7.3, or with 2.6.6: test passes with both.
So I need a full unit test to reproduce this, unless it might have been fixed in 2.6.4 or later.
